File: 208. Implement Trie (Prefix Tree).py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isEndofWord = False

class Trie:
    '''
    https://www.geeksforgeeks.org/trie-insert-and-search/
    Replace the list with dictionary
    '''
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self.root
        length = len(word)
        for level in range(length):
            pos = word[level]
            if pos not in node.child:
                node.child[pos] = TrieNode()
            node = node.child[pos]
        node.isEndofWord = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self.root
        length = len(word)
        for level in range(length):
            pos = word[level]
            if pos not in node.child:
                return False
            node = node.child[pos]
        if node.isEndofWord:
            return True
        else:
            return False

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.root
        length = len(prefix)
        for level in range(length):
            pos = prefix[level]
            if pos not in node.child:
                return False
            node = node.child[pos]
        return True

        
        
# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208. Implement Trie (Prefix Tree).java
class Trie {
    class TrieNode {
        boolean isEnd;
        TrieNode[] children;
        public TrieNode() {
            isEnd = false;
            children = new TrieNode[26];
        }
    }

    TrieNode root;
    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode curr = root;
        for (int i=0; i<word.length(); i++) {
            char currchar = word.charAt(i);
            if (curr.children[currchar-'a'] == null) {
                curr.children[currchar-'a'] = new TrieNode();
            }
            curr = curr.children[currchar-'a'];
        }
        curr.isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode curr = root;
        for (int i=0; i<word.length(); i++) {
            char currchar = word.charAt(i);
            if (curr.children[currchar-'a'] == null) {
                return false;
            }
            curr = curr.children[currchar-'a'];
        }
        if (!curr.isEnd) {
            return false;
        }
        return true;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for (int i=0; i<prefix.length(); i++) {
            char currchar = prefix.charAt(i);
            if (curr.children[currchar-'a'] == null) {
                return false;
            }
            curr = curr.children[currchar-'a'];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */